Fixed #27792 -- Extended category support for feeds

PavloKapyshin opened the pull request:
[Ticket #27792](https://code.djangoproject.com/ticket/27792)

foresmac commented on docs/ref/utils.txt:
```diff
@@ -347,13 +347,25 @@ https://web.archive.org/web/20110718035220/http://diveintomark.org/archives/2004
+
+        .. versionchanged:: 2.0
+
+            Now it is possible to specify :class:`Category` instances
```
Instead of "Now it is possible", try "It is now possible".

foresmac commented on docs/ref/utils.txt:
```diff
@@ -347,13 +347,25 @@ https://web.archive.org/web/20110718035220/http://diveintomark.org/archives/2004
+        .. versionchanged:: 2.0
+
+            Now it is possible to specify :class:`Category` instances
+            in ``categories`` argument.
```
try "in the `categories` argument"

foresmac commented on docs/ref/utils.txt:
```diff
@@ -347,13 +347,25 @@ https://web.archive.org/web/20110718035220/http://diveintomark.org/archives/2004
+        .. versionchanged:: 2.0
+
+            Now it is possible to specify :class:`Category` instances
+            in ``categories`` argument.
```
See above

foresmac commented on docs/releases/2.0.txt:
```diff
@@ -102,7 +102,9 @@ Minor features
 
-* ...
+* Class :class:`~django.utils.feedgenerator.Category` allows to
+  specify categories with ``scheme`` (for Atom feeds), ``domain``
```
Instead of "allows to specify", try "allows specifying".

PavloKapyshin commented:
Thank you, @foresmac, I’ve updated pull request.

vytisb commented on django/utils/feedgenerator.py:
```diff
@@ -109,19 +109,29 @@ def to_str(s):
+        res = []
+        for cat in categories:
+            if not isinstance(cat, Category):
+                cat = force_text(cat, strings_only=True)
```
Is `string_only=True` required here? If it is, it's not tested.

PavloKapyshin commented:
Good point, @vytisb, thank you. I extended test feeds’ categories to check for use of `force_text`.

PavloKapyshin commented:
I’ve updated pull request to keep up with master.

timgraham commented:
Considering all changes are in `utils/feedgenerator.py`, what do you think about putting the tests in `tests/utils_tests/test_feedgenerator.py` rather than in the syndication tests? It might be overcomplicated to go through the request/response cycle to test this.

timgraham commented on tests/syndication_tests/feeds.py:
```diff
@@ -53,6 +53,22 @@ class TestRss091Feed(TestRss2Feed):
+    item_categories = (
+        'python', 'testing', 11, 12,
+        feedgenerator.Category('simple', domain='http://example.org/Simple'),
+        feedgenerator.Category(b'creativity in tests', scheme=b'bytes here'),
```
I'm doubtful that there's a need for bytestring handling considering the sentiment in https://code.djangoproject.com/ticket/27795. Did you have a use case in mind? Could some `force_text()` usage be removed instead?

PavloKapyshin commented:
I thought about tests initially, and concluded that given this patch affects syndication and includes documentation additions related to it, it’d be beneficial to have tests for an API that is documented.

Replaced bytestrings with realistic example: translated strings. Replaced `force_text` with `str` (to keep patch focused, only where code is changed).

PavloKapyshin commented:
So, @timgraham, what do you think? Should I move tests to `test_feedgenerator.py`?

timgraham commented:
I think simpler tests are better. Existing tests might also be more complicated than needed. Testing something in django.utils exclusively in a a contrib app doesn't seem ideal. Some tests in syndication could be fine if there is some reason to suspect it's 'usage of FeedGenerator might affect things but testing all aspects of FeedGenerator via syndication feels more complicated than needed.

PavloKapyshin commented:
@timgraham, I’ve moved and simplified tests. Patch is updated. Unfortunately, in addition to that—to make `test_atom_support` work—I had (for now) to pass `enclosures=()` to `feed.add_item`.

[75cf9b](https://github.com/django/django/commit/75cf9b5ac031feb8f94271c9906157c921a14520) introduced a regression by removing `if enclosure is None:`: when `add_item` in Atom feed does not receive `enclosures` (that is, when its value is `None`), `TypeError` is raised (due to `for enclosure in item['enclosures']:` in `add_item_elements` of `Atom1Feed`).

When aforementioned regression will be fixed, I’ll update this pull request.

timgraham commented:
Thanks, are you able to fix that regression also?

PavloKapyshin commented:
Sure: #8110.

PavloKapyshin commented:
Thank you, @timgraham. I updated pull request.

PavloKapyshin commented:
Just wanted to say I did not give up: I rebase this on top of `master` every ~10 days :)

timgraham commented:
Pavlo, did you close this by accident?

PavloKapyshin commented:
Tim, that was deliberate action. I think feed generation (syndication) should live in separate specialized library, that is not tied to particular web framework or ORM.

timgraham commented:
Okay, so you would withdraw the ticket and advocate for removing contrib.syndication from Django (or at least not make further enhancements to it)?
